src/Symfony/Component/Uri/Uri.php

```diff
@@ -0,0 +1,172 @@
+    private const URI_AUTHORITY_REGEX = '/^(?:(?P<user>[^:@]*)(?::(?P<pass>[^@]*))?@)?(?P<host>[^:]*)(?::(?P<port>\d*))?$/';
+
+    public function __construct(
+        public string $scheme,
+        #[\SensitiveParameter]
+        public ?string $user = null,
+        #[\SensitiveParameter]
+        public ?string $password = null,
+        public ?string $host = null,
+        public ?int $port = null,
+        public ?string $path = null,
+        public ?QueryString $query = null,
+        public ?string $fragment = null,
+        public ?FragmentTextDirective $fragmentTextDirective = null,
+    ) {
+    }
+
+    /**
+     * Parses a URL.
+     *
+     * The `user` and `pass` keys are url-decoded automatically when parsing.
+     *
+     * @throws InvalidUriException
+     */
+    public static function parse(string $uri): static
+    {
+        preg_match(self::URI_GLOBAL_REGEX, $uri, $matches);
+        if (!$matches || !isset($matches['scheme']) || '' === $matches['scheme']) {
+            throw new InvalidUriException($uri);
+        }
+
+        if (isset($matches['authority'])) {
+            preg_match(self::URI_AUTHORITY_REGEX, $matches['authority'], $authMatches);
+
+            $matches = array_merge($matches, $authMatches);
+            unset($matches['authority']);
+        }
+
+        $matches = array_filter($matches, static fn (string $value): bool => '' !== $value);
+
+        return new static(
+            $matches['scheme'],
+            isset($matches['user']) ? rawurldecode($matches['user']) : null,
+            isset($matches['pass']) ? rawurldecode($matches['pass']) : null,
+            $matches['host'] ?? null,
+            $matches['port'] ?? null,
+            $matches['path'] ?? null,
+            isset($matches['query']) ? QueryString::parse($matches['query']) : null,
+            $matches['fragment'] ?? null,
+        );
+    }
+
+    /**
+     * Resolves a relative URI against a base URI.
+     *
+     * Uri::resolve('http://example.com', '/foo/bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', '../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '../../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', 'http://example.org/bar'); // http://example.org/bar
+     */
+    public static function resolve(self|string $baseUri, string $relativeUri): string
+    {
+        if ('' === $relativeUri) {
+            return (string) $baseUri;
+        }
+
+        // the relative URI is an absolute URI
+        if (str_contains($relativeUri, '://')) {
```

Comment:
This will fail for URIs such as `mailto:me@example.com`.

Reply:
Thank you for pointing this issue, I addressed this case in https://github.com/symfony/symfony/compare/a95f3840c2182f2617d0e7eff082a16e5c029c69..6dfdba2854d8defe621146be83acfaefc52d5494.